[InstCombine] Convert @log to @llvm.log if the input is known positive.

davemgreen opened the pull request:
Similar to 112aac4e8961b9626bb84f36deeaa5a674f03f5a, this converts log libcalls to llvm.log.f64 intrinsics if we know they do not set errno, as the input is not zero and not negative. As log will produce errno if the input is 0 (returning -inf) or if the input is negative (returning nan), we also perform the conversion when we have noinf and nonan.

davemgreen commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2797,22 +2813,21 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
   // frem.
   bool IsNoNan = CI->hasNoNaNs();
   if (!IsNoNan) {
+    SimplifyQuery SQ(DL, TLI, DT, AC, CI, true, true, DC);
```
This is unrelated, I'll push it separately.

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2589,10 +2584,26 @@ Value *LibCallSimplifier::optimizeLog(CallInst *Log, IRBuilderBase &B) {
+    if (IsKnownNoErrno) {
+      IRBuilderBase::FastMathFlagGuard Guard(B);
+      B.setFastMathFlags(Log->getFastMathFlags());
+      return B.CreateIntrinsic(LogID, {Ty}, {Log->getArgOperand(0)});
```
Losing metadata from the call (i.e. !fpmath)

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2589,10 +2584,26 @@ Value *LibCallSimplifier::optimizeLog(CallInst *Log, IRBuilderBase &B) {
+          Log->getOperand(0), KnownFPClass::OrderedLessThanZeroMask,
+          /*Depth=*/0, SQ);
+      IsKnownNoErrno =
+          Known.isKnownNeverZero() && Known.cannotBeOrderedLessThanZero();
```
Broken for DAZ, you want isKnownNeverLogicalZero (we should probably remove isKnownNeverZero)

davemgreen replied:
I had given denormals a test with log and in the routine I was using they had not been treated as zero (unlike fmod where they were). I've changed this over though.

arsenm replied:
It entirely depends on what happens in the implementation, you have to conservatively assume denormals may be treated as zeroes.

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2589,10 +2584,28 @@ Value *LibCallSimplifier::optimizeLog(CallInst *Log, IRBuilderBase &B) {
+          Log->getOperand(0),
+          KnownFPClass::OrderedLessThanZeroMask | fcSubnormal,
+          /*Depth=*/0, SQ);
+      Function *F = Log->getParent()->getParent();
+      IsKnownNoErrno = Known.isKnownNeverLogicalZero(*F, Ty) &&
+                       Known.cannotBeOrderedLessThanZero();
```
```suggestion
      Function *F = Log->getParent()->getParent();
      IsKnownNoErrno =  Known.cannotBeOrderedLessThanZero() &&
                       Known.isKnownNeverLogicalZero(*F, Ty);
```

arsenm commented on llvm/test/Transforms/InstCombine/log-to-intrinsic.ll:
```diff
@@ -0,0 +1,355 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
```
Fix to use the latest version of update_test_checks?

arsenm reviewed (approved):
lgtm with nits

davemgreen committed:
[InstCombine] Convert @log to @llvm.log if the input is known positive.

Similar to 112aac4e8961b9626bb84f36deeaa5a674f03f5a, this converts log libcalls
to llvm.log.f64 intrinsics if we know they do not set errno, as the input is
not zero and not negative. As log will produce errno if the input is 0
(returning -inf) or if the input is negative (returning nan), we also perform
the conversion when we have noinf and nonan.
